File: app/src/main/java/com/dimatest/movieapp/common/BaseUseCase.java
package com.dimatest.movieapp.common;

import androidx.core.util.Preconditions;
import androidx.lifecycle.MutableLiveData;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BaseUseCase {

    public MutableLiveData<Boolean> isLoading = new MutableLiveData<>();

    protected final Scheduler mainThread;
    protected final Scheduler ioThread;
    protected final CompositeDisposable disposables;

    public BaseUseCase(Scheduler mainThread, Scheduler ioThread) {
        this.mainThread = mainThread;
        this.ioThread = ioThread;
        this.disposables = new CompositeDisposable();
    }

    public void clearDisposable() {
        disposables.clear();
    }

    protected void addDisposable(Disposable disposable) {
        Preconditions.checkNotNull(disposable);
        Preconditions.checkNotNull(disposables);
        disposables.add(disposable);
    }
}

File: app/src/main/java/com/dimatest/movieapp/common/BaseUseCase.kt
package com.dimatest.movieapp.common

import androidx.lifecycle.MutableLiveData
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BaseUseCase(protected val mainThread: Scheduler, protected val ioThread: Scheduler) {

    var isLoading = MutableLiveData<Boolean>()
    private val disposables = CompositeDisposable()

    fun clearDisposable() {
        disposables.clear()
    }

    protected fun addDisposable(disposable: Disposable?) {
        disposable?.let { disposables.add(it) }
    }
}